Add action_msgs package

jacobperron opened the pull request:
Contains common message and service definitions for all ROS actions.

Messages:
- GoalStatus.msg - Contains a goal ID, timestamp, and status.
- GoalStatusArray.msg - An array of GoalStatus.msg.
- UUID.msg - Type of goal IDs; this is temporary until we port the existing uuid_msgs (see TODOs).

Service:
- CancelGoal.srv - API for canceling goals describing the cancel policy.

Resolves #40

jacobperron committed:
Add action_msgs package

Contains common message and service definitions for all ROS actions.

Messages:
- GoalStatus.msg - Contains a goal ID, timestamp, and status.
- GoalStatusArray.msg - An array of GoalStatus.msg.
- UUID.msg - Type of goal IDs; this is temporary until we port the existing uuid_msgs (see TODOs).

Service:
- CancelGoal.srv - API for canceling goals describing the cancel policy.

sloretz commented on action_msgs/package.xml:
```diff
@@ -0,0 +1,28 @@
+
+  <build_depend>builtin_interfaces</build_depend>
+
+  <exec_depend>builtin_interfaces</exec_depend>
```
I think `<depend>builtin_interfaces</depend>` to get all three `build_depend`, `exec_depend`, and `build_export_depend` would make sense because a C  generator would export a header with an include to `builtin_interfaces/msg/Time.h`

dirk-thomas replied:
In general there are only two types of dependencies: the ones for build time and the ones for run time. Internally `exec_depend` and `build_export_depend` do the same thing - they are only named separately since users found it "easier" / more intuitive than `run_depend`.

Anyway if you use both then using `depend` makes sense :+1:

jacobperron committed:
Add GoalInfo.msg encapsulating goal ID and timestamp

jacobperron committed:
Use <depend> tag to replace <build_depend> and <exec_depend>

jacobperron commented:
Ready for review.

I've added GoalInfo.msg for encapsulating the goal ID and timestamp and updated the package.xml to use a `<depend>` tag.

CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5304)](http://ci.ros2.org/job/ci_linux/5304/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2057)](http://ci.ros2.org/job/ci_linux-aarch64/2057/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4420)](http://ci.ros2.org/job/ci_osx/4420/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5283)](http://ci.ros2.org/job/ci_windows/5283/)

sloretz commented on action_msgs/srv/CancelGoal.srv:
```diff
@@ -0,0 +1,15 @@
+GoalInfo goal_info
+---
+# Goals that accepted the cancel request
+GoalInfo[] goals_canceling
```
I can't figure out why windows is warning about the cancel goal response. It doesn't give a line number. The only way I could see `uint8_t` being a type in that file is if something was setting one of the bytes of `UUID`

```
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.12.25827\include\xutility(2888): warning C4244: '=': conversion from 'const int' to 'uint8_t', possible loss of data (compiling source file C:\J\workspace\ci_windows\ws\build\action_msgs\rosidl_typesupport_fastrtps_cpp\action_msgs\srv\dds_fastrtps\cancel_goal__response__type_support.cpp) [C:\J\workspace\ci_windows\ws\build\action_msgs\action_msgs__rosidl_typesupport_fastrtps_cpp.vcxproj]
```

<details><summary>content of `build\action_msgs\rosidl_typesupport_fastrtps_cpp\action_msgs\srv\dds_fastrtps\cancel_goal__response__type_support.cpp`</summary>

```
// generated from rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em
// generated code does not contain a copyright notice

#include "action_msgs/srv/cancel_goal__response__rosidl_typesupport_fastrtps_cpp.hpp"

#include <limits>
#include <stdexcept>
#include <string>

#include "rosidl_typesupport_cpp/message_type_support.hpp"

#include "rosidl_typesupport_fastrtps_cpp/identifier.hpp"
#include "rosidl_typesupport_fastrtps_cpp/message_type_support.h"
#include "rosidl_typesupport_fastrtps_cpp/message_type_support_decl.hpp"

#include "fastcdr/Cdr.h"

// forward declaration of message dependencies and their conversion functions
namespace action_msgs
{
namespace msg
{
namespace typesupport_fastrtps_cpp
{
bool cdr_serialize(
  const action_msgs::msg::GoalInfo &,
  eprosima::fastcdr::Cdr &);
bool cdr_deserialize(
  eprosima::fastcdr::Cdr &,
  action_msgs::msg::GoalInfo &);
size_t get_serialized_size(
  const action_msgs::msg::GoalInfo &,
  size_t current_alignment);
size_t
max_serialized_size_GoalInfo(
  bool & full_bounded,
  size_t current_alignment);
}  // namespace typesupport_fastrtps_cpp
}  // namespace msg
}  // namespace action_msgs


namespace action_msgs
{

namespace srv
{

namespace typesupport_fastrtps_cpp
{

bool
ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC_action_msgs
cdr_serialize(
  const action_msgs::srv::CancelGoal_Response & ros_message,
  eprosima::fastcdr::Cdr & cdr)
{
  // field.name goals_canceling
  {
    size_t size = ros_message.goals_canceling.size();
    cdr << static_cast<uint32_t>(size);
    for (size_t i = 0; i < size; i++) {
      action_msgs::msg::typesupport_fastrtps_cpp::cdr_serialize(
        ros_message.goals_canceling[i],
        cdr);
    }
  }
  return true;
}

bool
ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC_action_msgs
cdr_deserialize(
  eprosima::fastcdr::Cdr & cdr,
  action_msgs::srv::CancelGoal_Response & ros_message)
{
  // field.name goals_canceling
  {
    uint32_t cdrSize;
    cdr >> cdrSize;
    size_t size = static_cast<size_t>(cdrSize);
    ros_message.goals_canceling.resize(size);
    for (size_t i = 0; i < size; i++) {
      action_msgs::msg::typesupport_fastrtps_cpp::cdr_deserialize(
        cdr, ros_message.goals_canceling[i]);
    }
  }

  return true;
}

size_t
ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC_action_msgs
get_serialized_size(
  const action_msgs::srv::CancelGoal_Response & ros_message,
  size_t current_alignment)
{
  size_t initial_alignment = current_alignment;

  const size_t padding = 4;
  (void)padding;

  // field.name goals_canceling
  {
    size_t array_size = ros_message.goals_canceling.size();

    current_alignment += padding +
      eprosima::fastcdr::Cdr::alignment(current_alignment, padding);

    for (size_t index = 0; index < array_size; ++index) {
      current_alignment +=
        action_msgs::msg::typesupport_fastrtps_cpp::get_serialized_size(
        ros_message.goals_canceling[index], current_alignment);
    }
  }

  return current_alignment - initial_alignment;
}

size_t
ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC_action_msgs
max_serialized_size_CancelGoal_Response(
  bool & full_bounded,
  size_t current_alignment)
{
  size_t initial_alignment = current_alignment;

  const size_t padding = 4;
  (void)padding;
  (void)full_bounded;

  // field.name goals_canceling
  {
    size_t array_size = 0;
    full_bounded = false;
    current_alignment += padding +
      eprosima::fastcdr::Cdr::alignment(current_alignment, padding);


    for (size_t index = 0; index < array_size; ++index) {
      current_alignment +=
        action_msgs::msg::typesupport_fastrtps_cpp::max_serialized_size_GoalInfo(
        full_bounded, current_alignment);
    }
  }

  return current_alignment - initial_alignment;
}

static bool __cdr_serialize(
  const void * untyped_ros_message,
  eprosima::fastcdr::Cdr & cdr)
{
  auto typed_message =
    static_cast<const action_msgs::srv::CancelGoal_Response *>(
    untyped_ros_message);
  return cdr_serialize(*typed_message, cdr);
}

static bool __cdr_deserialize(
  eprosima::fastcdr::Cdr & cdr,
  void * untyped_ros_message)
{
  auto typed_message =
    static_cast<action_msgs::srv::CancelGoal_Response *>(
    untyped_ros_message);
  return cdr_deserialize(cdr, *typed_message);
}

static uint32_t __get_serialized_size(
  const void * untyped_ros_message)
{
  auto typed_message =
    static_cast<const action_msgs::srv::CancelGoal_Response *>(
    untyped_ros_message);
  return static_cast<uint32_t>(get_serialized_size(*typed_message, 0));
}

static size_t __max_serialized_size(bool & full_bounded)
{
  return max_serialized_size_CancelGoal_Response(full_bounded, 0);
}

static message_type_support_callbacks_t callbacks = {
  "action_msgs",
  "CancelGoal_Response",
  __cdr_serialize,
  __cdr_deserialize,
  __get_serialized_size,
  __max_serialized_size
};

static rosidl_message_type_support_t handle = {
  rosidl_typesupport_fastrtps_cpp::typesupport_identifier,
  &callbacks,
  get_message_typesupport_handle_function,
};

}  // namespace typesupport_fastrtps_cpp

}  // namespace srv

}  // namespace action_msgs

namespace rosidl_typesupport_fastrtps_cpp
{

template<>
ROSIDL_TYPESUPPORT_FASTRTPS_CPP_EXPORT_action_msgs
const rosidl_message_type_support_t *
get_message_type_support_handle<action_msgs::srv::CancelGoal_Response>()
{
  return &action_msgs::srv::typesupport_fastrtps_cpp::handle;
}

}  // namespace rosidl_typesupport_fastrtps_cpp

#ifdef __cplusplus
extern "C"
{
#endif

const rosidl_message_type_support_t *
ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(rosidl_typesupport_fastrtps_cpp, action_msgs, srv, CancelGoal_Response)() {
  return &action_msgs::srv::typesupport_fastrtps_cpp::handle;
}

#ifdef __cplusplus
}
#endif

```
</details>

jacobperron commented:
CI (with the fix from ros2/rosidl#307):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5349)](http://ci.ros2.org/job/ci_linux/5349/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2079)](http://ci.ros2.org/job/ci_linux-aarch64/2079/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4442)](http://ci.ros2.org/job/ci_osx/4442/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5312)](http://ci.ros2.org/job/ci_windows/5312/)

jacobperron commented:
Blocked by ros2/rosidl#307

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5351)](http://ci.ros2.org/job/ci_linux/5351/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2081)](http://ci.ros2.org/job/ci_linux-aarch64/2081/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4444)](http://ci.ros2.org/job/ci_osx/4444/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5314)](http://ci.ros2.org/job/ci_windows/5314/)

sloretz commented on action_msgs/msg/GoalInfo.msg:
```diff
@@ -0,0 +1,6 @@
+# Goal ID
+# TODO(jacobperron): Use a ported version of https://github.com/ros-geographic-info/unique_identifier/blob/master/uuid_msgs
```
Given that `uuid_msgs` has just a single message, I think the UUID.msg could live in `builtin_interfaces` rather than porting a package named `uuid_msgs`. Porting the tools in `unique_id` wouldn't need to block that.

sloretz reviewed (approved):
LGTM. Just one opinion that UUID could live in builtin_interfaces, but that is a conversation that can be resolved later.

jacobperron commented:
> Just one opinion that UUID could live in builtin_interfaces, but that is a conversation that can be resolved later.

Sounds good.

dirk-thomas commented:
`builtin_interfaces` only contains types defined in DDS. I don't think we should add any other ROS specific types to it.

jacobperron commented:
@sloretz @dirk-thomas 
Created a ticket regarding UUID message definitions: ros2/ros2#593

dirk-thomas commented:
See ros2/common_interfaces#64.
